phidgets_high_speed_encoder/src/phidgets_high_speed_encoder.cpp

```diff
@@ -0,0 +1,250 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#include <ros/ros.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sstream>
+#include <libphidgets/phidget21.h>
+#include <std_msgs/String.h>
+#include "phidgets_high_speed_encoder/encoder_params.h"
+
+static CPhidgetEncoderHandle phid;
+static ros::Publisher encoder_pub;
+static bool initialised = false;
+static std::string frame_id;
+static bool inverted = false;
+
+int AttachHandler(CPhidgetHandle phid, void *userptr)
+{
+  int serial_number;
+  const char *name;
+
+  CPhidget_getDeviceName(phid, &name);
+  CPhidget_getSerialNumber(phid, &serial_number);
+  ROS_INFO("%s Serial number %d attached!",
+           name, serial_number);
+
+  return 0;
+}
+
+int DetachHandler(CPhidgetHandle phid, void *userptr)
+{
+  int serial_number;
+  const char *name;
+
+  CPhidget_getDeviceName(phid, &name);
+  CPhidget_getSerialNumber(phid, &serial_number);
+  ROS_INFO("%s Serial number %d detached!",
+           name, serial_number);
+
+  return 0;
+}
+
+int ErrorHandler(CPhidgetHandle phid, void *userptr,
+                 int ErrorCode, const char *Description)
+{
+  ROS_INFO("Error handled. %d - %s", ErrorCode, Description);
+  return 0;
+}
+
+int InputChangeHandler(CPhidgetEncoderHandle ENC,
+                       void *usrptr, int Index, int State)
+{
+  return 0;
+}
+
+int PositionChangeHandler(CPhidgetEncoderHandle ENC,
+                          void *usrptr, int Index,
+                          int Time, int RelativePosition)
+{
+  static uint32_t sequence_number = 0U;
+  int Position;
+  CPhidgetEncoder_getPosition(ENC, Index, &Position);
+
+  phidgets_high_speed_encoder::encoder_params e;
+  e.header.stamp = ros::Time::now();
+  e.header.frame_id = frame_id;
+  e.header.seq = sequence_number++;
+  e.index = Index;
+  e.count = (inverted ? -Position : Position);
+  e.count_change = (inverted ? -RelativePosition : RelativePosition);
+  e.time = Time;
```

Comment:
The IMU is a little different, because it publishes periodically. The encoder only sends message, when there is at least one tick.

I noticed that the phidgets IMU resynchronizes if it is too old: https://github.com/ccny-ros-pkg/phidgets_drivers/blob/master/phidgets_imu/src/imu_ros_i.cpp#L180. I'm not sure if there is still value in using the devices time, when it resynchronizes often.